Packet++/src/LdapLayer.cpp

```diff
@@ -0,0 +1,210 @@
+	{
+		Asn1IntegerRecord messageIdRecord(messageId);
+		Asn1ConstructedRecord messageRootRecord(Asn1TagClass::Application, operationType, messageRecords);
+
+		std::vector<Asn1Record*> rootSubRecords = {&messageIdRecord, &messageRootRecord};
+
+		std::unique_ptr<Asn1ConstructedRecord> controlsRecord;
+		if (!controls.empty())
+		{
+			PointerVector<Asn1Record> controlsSubRecords;
+			for (const auto& control : controls)
+			{
+				Asn1OctetStringRecord controlTypeRecord(control.controlType);
+				if (control.controlValue.empty())
+				{
+					controlsSubRecords.pushBack(new Asn1SequenceRecord({&controlTypeRecord}));
+				}
+				else
+				{
+					auto controlValueSize = static_cast<size_t>(control.controlValue.size() / 2);
+					std::unique_ptr<uint8_t[]> controlValue(new uint8_t[controlValueSize]);
+					controlValueSize = hexStringToByteArray(control.controlValue, controlValue.get(), controlValueSize);
+					Asn1OctetStringRecord controlValueRecord(controlValue.get(), controlValueSize);
+					controlsSubRecords.pushBack(new Asn1SequenceRecord({&controlTypeRecord, &controlValueRecord}));
+				}
+			}
+			controlsRecord = std::unique_ptr<Asn1ConstructedRecord>(new Asn1ConstructedRecord(Asn1TagClass::ContextSpecific, 0, controlsSubRecords));
+			rootSubRecords.push_back(controlsRecord.get());
+		}
+
+		Asn1SequenceRecord rootRecord(rootSubRecords);
+
+		auto encodedData = rootRecord.encode();
+		m_DataLen = encodedData.size();
+		m_Data = new uint8_t[m_DataLen];
+		std::copy(encodedData.begin(), encodedData.end(), m_Data);
+		m_Protocol = LDAP;
+		m_Asn1Record = Asn1Record::decode(m_Data, m_DataLen, true);
+	}
+
+	std::string LdapLayer::toString() const
+	{
+		auto extendedInfo = getExtendedStringInfo();
+		return "LDAP Layer, " + getLdapOperationType().toString() + (extendedInfo.empty() ? "" : ", " + extendedInfo);
+	}
+
+	LdapLayer* LdapLayer::parseLdapMessage(uint8_t* data, size_t dataLen, Layer* prevLayer, Packet* packet)
+	{
+		try
+		{
+			auto asn1Record = Asn1Record::decode(data, dataLen, true);
+			auto operationType = LdapOperationType::fromIntValue(asn1Record->castAs<Asn1SequenceRecord>()->getSubRecords().at(1)->getTagType());
+			if (operationType != LdapOperationType::Unknown)
+			{
+				return new LdapLayer(asn1Record, data, dataLen, prevLayer, packet);
+			}
+
+			return nullptr;
+		}
+		catch (...)
+		{
+			return nullptr;
+		}
+	}
+
+	Asn1SequenceRecord* LdapLayer::getRootAsn1Record() const
+	{
+		return m_Asn1Record->castAs<Asn1SequenceRecord>();
+	}
+
+	Asn1ConstructedRecord* LdapLayer::getLdapOperationAsn1Record() const
+	{
+		return getRootAsn1Record()->getSubRecords().at(1)->castAs<Asn1ConstructedRecord>();
+	}
+
+	uint16_t LdapLayer::getMessageID() const
+	{
+		return getRootAsn1Record()->getSubRecords().at(0)->castAs<Asn1IntegerRecord>()->getValue();
+	}
+
+	std::vector<LdapControl> LdapLayer::getControls() const
+	{
+		std::vector<LdapControl> controls;
+		if (getRootAsn1Record()->getSubRecords().size() < 3)
```

Comment:
is 3 a magic number? can we have a comment on it?

Reply:
also for the following number `2`, and `1`.

Reply:
This is the LDAP specification that you can find here: https://github.com/vlm/ldap-server-example/blob/master/Lightweight-Directory-Access-Protocol-V3.asn1

```
        LDAPMessage ::= SEQUENCE {
             messageID       MessageID,
             protocolOp      CHOICE {
                  bindRequest           BindRequest,
                  bindResponse          BindResponse,
                  unbindRequest         UnbindRequest,
                  searchRequest         SearchRequest,
                  searchResEntry        SearchResultEntry,
                  searchResDone         SearchResultDone,
                  searchResRef          SearchResultReference,
                  modifyRequest         ModifyRequest,
                  modifyResponse        ModifyResponse,
                  addRequest            AddRequest,
                  addResponse           AddResponse,
                  delRequest            DelRequest,
                  delResponse           DelResponse,
                  modDNRequest          ModifyDNRequest,
                  modDNResponse         ModifyDNResponse,
                  compareRequest        CompareRequest,
                  compareResponse       CompareResponse,
                  abandonRequest        AbandonRequest,
                  extendedReq           ExtendedRequest,
                  extendedResp          ExtendedResponse,
                  ...,
                  intermediateResponse  IntermediateResponse },
             controls       [0] Controls OPTIONAL }
```

- Message ID is a mandatory first parameter
- LDAP operation type is a mandatory second parameter
- Controls is an optional third parameter

Reply:
I would suggest using a meaningful `constexpr` variable for better understanding the code.
e.g. `constexpr int messageIDindex = 1`

Reply:
That's a good idea! Added in this commit: https://github.com/seladb/PcapPlusPlus/pull/1412/commits/7a7480f4f49dedb4e3361d45015e6dd71368d2b1